app/src/test/java/org/sagebionetworks/bridge/researchstack/CrfDataProviderTest.java

```diff
@@ -0,0 +1,514 @@
+        when(bridgeManagerProvider.getConsentDao()).thenReturn(consentDAO);
+        when(bridgeManagerProvider.getAuthenticationManager()).thenReturn(authenticationManager);
+        when(bridgeManagerProvider.getActivityManager()).thenReturn(activityManager);
+
+        pinCodeConfig = mock(PinCodeConfig.class);
+        fileAccess = mock(FileAccess.class);
+        appDatabase = mock(BridgeEncryptedDatabase.class);
+
+        storageAccess = mock(StorageAccessWrapper.class);
+        when(storageAccess.getPinCodeConfig()).thenReturn(pinCodeConfig);
+        when(storageAccess.getAppDatabase()).thenReturn(appDatabase);
+        when(storageAccess.getFileAccess()).thenReturn(fileAccess);
+
+        when(apiClientProvider.getClient(AuthenticationApi.class)).thenReturn(authenticationApi);
+        when(apiClientProvider
+                .getClient(same(ForConsentedUsersApi.class), any(SignIn.class)))
+                .thenReturn(forConsentedUsersApi);
+
+        resetTest(null, true, null);
+
+        PowerMockito.mockStatic(PreferenceManager.class);
+        PowerMockito.mockStatic(Looper.class);
+
+        AppPrefs.init(context);
+    }
+
+    private void resetTest(DateTime firstSignInDate, boolean assignClinic1, String initialDataGroup) {
+        dataProvider = new MockCrfDataProvider(
+                firstSignInDate, assignClinic1, initialDataGroup,
+                researchStackDAO, storageAccess, taskHelper);
+        dataProvider.mockGetAllActiities(successList(assignClinic1));
+    }
+
+    @Test
+    public void testInitialize() {
+        dataProvider.initialize(context).test().assertCompleted();
+    }
+
+    @Test
+    public void testFindClinc1() {
+        final ScheduledActivityListV4 successList = successList(true);
+        final ScheduledActivity clinic = dataProvider.findActivity(successList, "clinic1");
+        assertNotNull(clinic);
+        assertEquals("clinic1", clinic.getActivity().getTask().getIdentifier());
+    }
+
+    @Test
+    public void testFindClinc2() {
+        final ScheduledActivityListV4 successList = successList(true);
+        final ScheduledActivity clinic = dataProvider.findActivity(successList, "clinic2");
+        assertNotNull(clinic);
+        assertEquals("clinic2", clinic.getActivity().getTask().getIdentifier());
+    }
+
+    @Test
+    public void testFindFail() {
+        final ScheduledActivityListV4 successList = successList(true);
+        final ScheduledActivity clinic = dataProvider.findActivity(successList, "clinic3");
+        assertNull(clinic);
+    }
+
+    @Test
+    public void testAssignClinic1() {
+        resetTest(null, true, null);
+        final ScheduledActivityListV4 successList = successList(true);
+        dataProvider.mockGetAllActiities(successList);
+
+        dataProvider.mockGetClinicActiities(clinicActivities(null, null));
+        dataProvider.getCrfActivities(new CrfDataProvider.CrfActivitiesListener() {
```

Comment:
Is it worth adding a check to verify that we correctly update the participant with the data group, and that we correctly update the clinic1 schedule to set the Finished On? Should we also have the user start with another data group to make sure we preserve existing data groups?